### DISCUSSION TITLE
Fastest way to override gutenberg components.

### DISCUSSION THREAD

**Discussion by ritchiemarkov** [gregsullivan/_tw#83] [Q&A] [answered]
Hello Greg and dear community.

My question is: in order to overide gutenberg default elements adding my tailwind / alpine classes should i use filter php hook, or there is a better way to make it happen? 

Theme is rocking, but i found (maybe its just me) there is a lot of stuff to do in order to work properly with gutenberg (fking gutenberg ok).

For example adding dark mode classes, colors, fonts.

**Comment by gregsullivan** [comment]
Hi Richard—Thanks for posting this!

Which elements do you have in mind? It sounds like you may also be looking to add Alpine.js directives; is that right?

If you're just looking to modify the Tailwind Typography output within Gutenberg, you'll want to edit the element modifiers towards the top of `functions.php`. (For example, you could add `dark:prose-invert`, and that might get you what you're looking for in terms of dark mode support, though you'll need to tweak the link colour, etc.

Hope that helps! I'd be happy to expand on the above with a better idea of what you're hoping to change.

**Comment by ritchiemarkov** [comment]
Hello! Let's suppose I wanna my h1 elements be yellow and green bold in
dark mode for an example. Should I use plain php for it or theme somehow
manage it? I just don't understand this way of manage Gutenberg and
tailwind.

Can you please explain me.

Il Lun 27 Nov 2023, 20:45 Greg Sullivan ***@***.***> ha
scritto:

**Reply by gregsullivan** [reply, marked as answer]
This part of the Tailwind Typography documentation should give you most of what you need:

https://tailwindcss.com/docs/typography-plugin#element-modifiers

So if you want them yellow, you'd add this:

`prose-h1:text-yellow`

If you wanted green and bold in dark mode, you'd add this:

`dark:prose-h1:text-green dark:prose-h1:font-bold`

(The above assumes you have colours set for `yellow` and `green`.)

In terms of where to put these in _tw, there's a constant for Tailwind Typography class modifiers towards the top of `functions.php`.

Hope that helps!

**Reply by ritchiemarkov** [reply]
ok i think i messed up something in my theme, with layouts, so typography isnt working. 

I mean its not by default apply on my text generated by gutenberg.
So im gonna investigate.

For now big thanks for your time.

**Reply by gregsullivan** [reply]
Sorry, I did just realize something: The `h1` isn't actually part of the block editor's post contents, and Tailwind Typography only applies to the post contents. So my example above would work for `h2`–`h6`, but not for `h1`.

For `h1`, you can modify post titles in the `components.css` file in `./tailwind/custom/components`.

**Reply by ritchiemarkov** [reply]
your theme should add by default right? 

i mean even if i put my content into 
<div class="contenuti mt-12 prose">
        <?php the_content(); ?>
      </div>
      
     
     
     
     and in my functions.php as you mentioned: 
     
     define(
		'NORD_TYPOGRAPHY_CLASSES',
		'prose prose-neutral max-w-none prose-a:text-yellow',
        'dark:prose-h2:text-green dark:prose-h2:font-bold'
	);
}



i cant see any effect, or my <div class="contenuti mt-12 prose"> affects it?

**Reply by ritchiemarkov** [reply]
`div class="contenuti mt-12 prose">
        <?php the_content(); ?>
      </div>`

**Reply by gregsullivan** [reply]
You'll need to keep the `_tw_content_class` (probably `nord_content_class` given your theme prefix) function as seen here:

```php
<div <?php _tw_content_class( 'entry-content' ); ?>>
```

You could change the above to:

```php
<div <?php _tw_content_class( 'contenuti mt-12' ); ?>>
```

**Reply by ritchiemarkov** [reply]
ok greg i swear this time its my last mesage for you. 

let's suppose my template is: 

`<?php
/*
	 * Template Name: Typos
	 * Template Post Type: post
	 */

get_header();  ?>

<main>

      
      <div>
        <?php the_content(); ?>
      </div>
      
</main>

  


<?php get_footer(); ?>`


and my functions.php is `?php
/**
 * nord functions and definitions
 *
 * @link https://developer.wordpress.org/themes/basics/theme-functions/
 *
 * @package nord
 */

if ( ! defined( 'NORD_VERSION' ) ) {
	/*
	 * Set the theme’s version number.
	 *
	 * This is used primarily for cache busting. If you use `npm run bundle`
	 * to create your production build, the value below will be replaced in the
	 * generated zip file with a timestamp, converted to base 36.
	 */
	define( 'NORD_VERSION', '0.1.0' );
}

if ( ! defined( 'NORD_TYPOGRAPHY_CLASSES' ) ) {
	/*
	 * Set Tailwind Typography classes for the front end, block editor and
	 * classic editor using the constant below.
	 *
	 * For the front end, these classes are added by the `nord_content_class`
	 * function. You will see that function used everywhere an `entry-content`
	 * or `page-content` class has been added to a wrapper element.
	 *
	 * For the block editor, these classes are converted to a JavaScript array
	 * and then used by the `./javascript/block-editor.js` file, which adds
	 * them to the appropriate elements in the block editor (and adds them
	 * again when they’re removed.)
	 *
	 * For the classic editor (and anything using TinyMCE, like Advanced Custom
	 * Fields), these classes are added to TinyMCE’s body class when it
	 * initializes.
	 */
	define(
		'NORD_TYPOGRAPHY_CLASSES',
		'prose prose-neutral max-w-none prose-a:text-yellow',
        'dark:prose-h2:text-green dark:prose-h2:font-bold'
	);
}

if ( ! function_exists( 'nord_setup' ) ) :
	/**
	 * Sets up theme defaults and registers support for various WordPress features.
	 *
	 * Note that this function is hooked into the after_setup_theme hook, which
	 * runs before the init hook. The init hook is too late for some features, such
	 * as indicating support for post thumbnails.
	 */
	function nord_setup() {`
	
	
	
	what should i apply? 
	
	Im so sorry, i just need to get it once. 
	
	And yes before opening this ticket i have looked into theme docs.
	
	just one example and much much love from me

**Comment by gregsullivan** [comment]
In your template, change this:

```php
<!-- Content -->
  <div>
    <?php the_content(); ?>
  </div>
```

to this:

```php
<!-- Content -->
  <div <?php _tw_content_class( 'entry-content' ); ?>>
    <?php the_content(); ?>
  </div>
```

Meanwhile, when you define the constant, this won't work:

```php
define(
'NORD_TYPOGRAPHY_CLASSES',
'prose prose-neutral max-w-none prose-a:text-yellow',
'dark:prose-h2:text-green dark:prose-h2:font-bold'
);
```

You want this:

```php
define(
'NORD_TYPOGRAPHY_CLASSES',
'prose prose-neutral max-w-none prose-a:text-yellow dark:prose-h2:text-green dark:prose-h2:font-bold'
);
```

(The value for the constant needs to be in a single argument.)

**Reply by ritchiemarkov** [reply]
oh yeah! Woah its working!! incredible!

i can also apply to another elements and not only on content right?

woah now we talking! 

Thanl you a lot!!!! <3 <3 

Great theme, and thank you for being available

**Reply by gregsullivan** [reply]
Are you talking specifically about the `dark` variant? You can definitely use that anywhere you want! You can also used the `_tw_content_class` function (but with your function prefix) anywhere you want Tailwind Typography styles to be applied.

You're very welcome—happy to help!

**Comment by ritchiemarkov** [comment]
Yes. It's really awesome!

I was ending to make every single wp element by adding classes by php
filter. It was working well actually. But now I can make it quickly.


I can consider filter  approach in case I need some particular js library
classes by added by default.
Or some custom styling of wp elements.

I really enjoy using your theme.

My next step is to understand benefits of esbuild over manual js enqueveu.

Thanks once again. 100% solved.
Have a great week.

Il Lun 27 Nov 2023, 23:28 Greg Sullivan ***@***.***> ha
scritto:

**Reply by gregsullivan** [reply]
Great, I'm very happy to hear it!

As far as esbuild goes, the biggest advantage for me is being able to use npm to manage libraries, and to be able to follow, e.g., the ["As a module" directions](https://alpinejs.dev/essentials/installation#as-a-module) for something like Alpine.js.
